CPU load for Mac. Linux and windows needs testing.

kamenitxan opened the pull request:
https://github.com/dblock/oshi/issues/14

dblock commented:
The build needs to pass, to start with. I can check the OSX version after that.

dblock commented on src/main/java/oshi/util/ExecutingCommand.java:
```diff
@@ -35,6 +33,7 @@
 		} catch (IOException e) {
 			return null;
 		}
+		p.destroy();
```
Can you explain this?

kamenitxan replied:
removed p.waitFor(); so it works with top command. 
p.destroy(); just to be safe, that process is dead.

dblock replied:
I don't think you can do this. The process spawns and then just hangs there and you don't stop it. You need to find a  way to run something that returns. Maybe using `top` isn't the best way? On Linux I believe /proc/stat can give you a lot of this, no idea about OSX.

dblock commented on src/test/java/oshi/SystemInfoTest.java:
```diff
@@ -46,6 +46,13 @@ public void testGetMemory() {
+    public void testCpuLoad() {
+        SystemInfo si = new SystemInfo();
+        HardwareAbstractionLayer hal = si.getHardware();
+        System.out.println(hal.getProcessorLoad());
```
This isn't a test, it just prints it out. What we want is something that at least checks that the output is a non-zero % number for example.

dblock commented:
Please also update CHANGELOG and README.

dblock commented:
Now, stepping back from the implementation above, getProcessorLoad is too specific, especially if it returns a number. You want to create a structure that returns a lot of different load-related information that we can extend in the future. Top returns load averages, cpu usage for user, system, idle, etc. So something that reflects that closer.

kamenitxan commented:
Here are changes: https://github.com/kamenitxan/oshi/commit/5a412bd8eb54f8d4cba8c6a1ae9c7edda234a04f
Tests are ok and build too.

What about extending Processor interface? Then this would be called like this
 ```hal.getProcessors()[0].getProcessorLoad() ```

Also CentralProcessor classes implements Processor, but CentralProcessor has more methods. Shouldn't they be in Processor too?

dblock commented:
Travis runs Linux only, we have to have an implementation, can't have CI go red, so a Linux implementation is needed for anything to be merged.

I think it's a good idea to revisit these classes.

dblock commented on CHANGELOG.md:
```diff
@@ -9,6 +9,10 @@ Next
 
+1.2 (2/1/2014)
+================
+* Added support for CPU load in OSX
```
This should be under Next and should just be a bullet with your name (same format as above please).

dblock commented on src/main/java/oshi/util/ExecutingCommand.java:
```diff
@@ -21,7 +21,7 @@
 			return null;
 		} catch (InterruptedException e) {
-			return null;
+			e.printStackTrace();
```
This is not necessary.

kamenitxan commented:
I will get Linux and fix it there tomorow.

I will also add methods from CentralProcessor to Processor interface.

kamenitxan commented:
Fixed and added.

Before I add cpuload to Processor interface I think, it is also good idea to change hal interface.
Now it would be ```hal.getProcessors()[0].getLoad()```

I would like to make ```hal.getProcessor``` which returns first CPU (getProcessors()[0])
Then it would be ```hal.getProcessor().getLoad()```

dblock commented:
Why would you want a single processor when you have multiple? 

If CPU load is a combination on top of multiple processors, then it should probably not be under getProcessor, no?

dblock commented:
The code here is much better. I don't think I want an explicit `getProcessorLoad` that returns a float, so lets write a more sensible interface and this will become mergeable.

kamenitxan commented:
Because I have 1 CPU with more cores. 
I moved  ```getLoad() ``` to Processor interface -> ```hal.getProcessors()[0].getLoad();```
Someone could implement getLoad for cores, but I think it's not so needed.

dblock commented:
I like this, merging.

dblock commented:
I just noticed, is the README still correct? Shouldn't the load be per CPU? Please feel free to PR a fix if I am right.

kamenitxan commented:
Readme is correct. Load is for whole CPU, not for single CPU core.

dblock commented:
:+1:
